alphacsc/_d_solver.py

```diff
@@ -183,122 +227,144 @@ def get_max_error_dict(self, z_encoder):
         [Yellin2017] BLOOD CELL DETECTION AND COUNTING IN HOLOGRAPHIC LENS-FREE
         IMAGING BY CONVOLUTIONAL SPARSE DICTIONARY LEARNING AND CODING.
         """
+        assert z_encoder.n_channels == self.n_channels
 
-        raise NotImplementedError()
+        d0 = z_encoder.get_max_error_patch()
 
-    def _init_windower(self, z_encoder):
-        raise NotImplementedError()
+        d0 = self._windower.window(d0)
 
-    def _get_objective(self, z_encoder):
+        return self.prox(d0)
 
-        def objective(D, full=False):
+    def init_dictionary(self, X):
+        """Returns a dictionary for the signal X depending on D_init value.
 
-            D = self.windower.window(D)
+        Parameter
+        ---------
+        X: array, shape (n_trials, n_channels, n_times)
+            The data on which to perform CSC.
 
-            return z_encoder.compute_objective(D)
+        Return
+        ------
+        D : array shape (n_atoms, n_channels + n_times_atom) or
+                  shape (n_atoms, n_channels, n_times_atom)
+            The initial atoms to learn from the data.
+        """
 
-        return objective
+        D_hat = self.dict_strategy.initialize(X)
 
+        if not hasattr(self.dict_strategy, 'D_init'):
+            D_hat = self._windower.window(D_hat)
```

Comment:
Not sure what is the rational here, can you add a comment?

Reply:
It checks if `D_init` was an ndarray actually. The same check in `init_dictionary` as `if window and not isinstance(D_init, np.ndarray):`

I've put a comment, let me know if it is ok.